### PR TITLE
feat : upgraded the app with evm module to SDK v0.50 

### PR DESCRIPTION
**Pull request by vishal-kanna**

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -8,6 +8,21 @@ import (
 	"path/filepath"
 	"strings"
 
+	// "github.com/CosmWasm/wasmd/x/wasm"
+	// wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+
+	"cosmossdk.io/log"
+	// "github.com/CosmWasm/wasmd/x/wasm"
```
Comment (byte-bandit) [inline app/app.go:15]: What's up with the commented out imports?

Comment (deepan95dev) [reply app/app.go]: I pushed wasm fixes by uncommenting them in #1039

File: proto/palomachain/paloma/valset/snapshot.proto [hunk 2]
```diff
@@ -11,7 +11,7 @@ option go_package = "github.com/palomachain/paloma/x/valset/types";
 
 message Validator {
   bytes shareCount = 1 [
-    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.customtype) = "cosmossdk.io/math.Int",
```
Comment (aleem1314) [inline proto/palomachain/paloma/valset/snapshot.proto:14]: Why valset proto files updated in this PR?

File: x/evm/genesis_test.go [hunk 3]
```diff
@@ -45,9 +44,7 @@ var _ = g.Describe("genesis", func() {
 	g.BeforeEach(func() {
 		t := g.GinkgoT()
 		a = app.NewTestApp(t, false)
-		ctx = a.NewContext(false, tmproto.Header{
-			Height: 5,
-		})
+		ctx = a.NewContext(false)
```
Comment (aleem1314) [inline x/evm/genesis_test.go:47]: Where are we setting starting to 5?

File: x/evm/keeper/attest.go [hunk 4]
```diff
@@ -77,33 +77,33 @@ func (k Keeper) attestRouter(ctx sdk.Context, q consensus.Queuer, msg consensust
...
-		logger.WithError(err).Error("failed to cast to consensus message")
+		logger.With(err).Error("failed to cast to consensus message")
 		return err
 	}
 
 	evidence, err := k.findEvidenceThatWon(ctx, msg.GetEvidence())
 	if err != nil {
 		if errors.Is(err, ErrConsensusNotAchieved) {
-			logger.WithError(err).Error("consensus not achieved")
+			logger.With(err).Error("consensus not achieved")
```
Comment (aleem1314) [inline x/evm/keeper/attest.go:95]: Why this change? Please revert this.

File: x/evm/keeper/attest.go [hunk 5]
```diff
@@ -77,33 +77,33 @@ func (k Keeper) attestRouter(ctx sdk.Context, q consensus.Queuer, msg consensust
...
 			return nil
 		}
-		logger.WithError(err).Error("failed to find evidence")
+		logger.With(err).Error("failed to find evidence")
```
Comment (aleem1314) [inline x/evm/keeper/attest.go:98]: ditto

File: x/evm/keeper/attest.go [hunk 6]
```diff
@@ -120,11 +120,11 @@ func (k Keeper) attestRouter(ctx sdk.Context, q consensus.Queuer, msg consensust
 	// If we got up to here it means that the enough evidence was provided
 	actionMsg := consensusMsg.(*types.Message).GetAction()
 	_, chainReferenceID := q.ChainInfo()
-	logger = logger.WithFields("chain-reference-id", chainReferenceID)
+	logger = logger.With("chain-reference-id", chainReferenceID)
```
Comment (aleem1314) [inline x/evm/keeper/attest.go:123]: ditto. This change is not required for the SDK version update.

File: x/evm/keeper/attest.go [hunk 7]
```diff
@@ -478,11 +478,11 @@ func (k Keeper) initiateERC20TokenOwnershipTransfer(
 		return err
 	}
 
-	liblog.FromSDKLogger(k.Logger(ctx)).WithFields("new-message-id", msgID).Debug("initiateERC20TokenOwnershipTransfer triggered")
+	k.Logger(ctx).With("new-message-id", msgID).Debug("initiateERC20TokenOwnershipTransfer triggered")
```
Comment (aleem1314) [inline x/evm/keeper/attest.go:481]: ditto

File: x/evm/keeper/keeper.go [hunk 8]
```diff
@@ -146,8 +148,9 @@ func (k Keeper) PickValidatorForMessage(ctx sdk.Context, chainReferenceID string
 	return k.msgAssigner.PickValidatorForMessage(ctx, weights, chainReferenceID, requirements)
 }
 
-func (k Keeper) Logger(ctx sdk.Context) liblog.Logr {
-	return liblog.FromSDKLogger(ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName)))
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
```
Comment (aleem1314) [inline x/evm/keeper/keeper.go:151]: ditto

File: x/evm/module.go [hunk 9]
```diff
@@ -19,14 +21,17 @@ import (
 )
 
 var (
-	_ module.AppModule      = AppModule{}
+	// _ module.AppModule      = AppModule{}
```
Comment (aleem1314) [inline x/evm/module.go:24]: Ditto here. Why valset proto files updated in this PR?

File: x/valset/module.go [hunk 10]
```diff
@@ -19,7 +19,7 @@ import (
 )
 
 var (
-	_ module.AppModule      = AppModule{}
+	// _ module.AppModule      = AppModule{}
```
Comment (aleem1314) [inline x/valset/module.go:22]: Ditto here. Why valset proto files updated in this PR?

File: app/app.go [hunk 11]
```diff
@@ -336,6 +335,10 @@ func New(
 	std.RegisterInterfaces(interfaceRegistry)
 	bApp := baseapp.NewBaseApp(Name, logger, db, txConfig.TxDecoder(), baseAppOptions...)
 	bApp.SetTxEncoder(txConfig.TxEncoder())
+
+	std.RegisterLegacyAminoCodec(legacyAmino)
+	std.RegisterInterfaces(interfaceRegistry)
+	bApp.SetTxEncoder(txConfig.TxEncoder())
```
Comment (aleem1314) [inline app/app.go:341]:
```suggestion
```

File: proto/palomachain/paloma/valset/snapshot.proto (hunk 2 above)
Comment (vishal-kanna) [reply proto/palomachain/paloma/valset/snapshot.proto:14]: @aleem1314  because we are using it the evm module.

File: app/app.go [hunk 12]
```diff
@@ -229,6 +225,9 @@ var (
 		icatypes.ModuleName:         nil,
 		// wasm.ModuleName:                {authtypes.Burner},
 		// treasurymoduletypes.ModuleName: {authtypes.Burner, authtypes.Minter},
+		// gravitymoduletypes.ModuleName:  {authtypes.Minter, authtypes.Burner},
+		// wasm.ModuleName:                {authtypes.Burner},
+		// treasurymoduletypes.ModuleName: {authtypes.Burner, authtypes.Minter},
```
Comment (aleem1314) [inline app/app.go:230]:
```suggestion
```

Same here. This change is not related to this PR.

File: app/app.go [hunk 13]
```diff
@@ -419,6 +424,8 @@ func New(
 	scopedTransferKeeper := app.CapabilityKeeper.ScopeToModule(ibctransfertypes.ModuleName)
 	// scopedWasmKeeper := app.CapabilityKeeper.ScopeToModule(wasm.ModuleName)
 	// scopedConsensusKeeper := app.CapabilityKeeper.ScopeToModule(consensusmoduletypes.ModuleName)
+	// scopedWasmKeeper := app.CapabilityKeeper.ScopeToModule(wasm.ModuleName)
+	// scopedConsensusKeeper := app.CapabilityKeeper.ScopeToModule(consensusmoduletypes.ModuleName)
```
Comment (aleem1314) [inline app/app.go:428]:
```suggestion
```

File: app/app.go [hunk 14]
```diff
@@ -515,7 +522,7 @@ func New(
 	app.IBCFeeKeeper = ibcfeekeeper.NewKeeper(
 		appCodec,
 		keys[ibcfeetypes.StoreKey],
-		app.IBCKeeper.ChannelKeeper, // may be replaced with IBC middleware
+		app.IBCKeeper.ChannelKeeper,
```
Comment (aleem1314) [inline app/app.go:525]:
```suggestion
		app.IBCKeeper.ChannelKeeper,  // may be replaced with IBC middleware
```

File: x/evm/keeper/attest.go (hunk 4 above)
Comment (vishal-kanna) [reply x/evm/keeper/attest.go:95]: Reverted the change and made the changes in the  util/liblog as well.

File: x/evm/genesis_test.go (hunk 3 above)
Comment (vishal-kanna) [reply x/evm/genesis_test.go]: NewContext function takes only one parameter i.e bool  and in NewContext function it calls the  other  function with the empty header.
In order to set the starting to 5, we need to use NewUncachedContext .

**Comment by aleem1314** [comment]
Here are few things we have to do 
- Params are deprecated. Move params to module state [Ref](https://github.com/cosmos/cosmos-sdk/issues/10514).
- Update module to use collections
- Add AutoCLI support

File: util/liblog/liblog.go [hunk 15]
```diff
@@ -11,13 +11,18 @@ type Logr interface {
...
 
 type lgwr struct {
 	l log.Logger
 }
 
+// Impl implements Logr.
+func (*lgwr) Impl() any {
+	return log.NewNopLogger()
+}
+
```
Comment (byte-bandit) [inline util/liblog/liblog.go:25]: Unneeded.

File: proto/palomachain/paloma/evm/params.proto [hunk 16]
```diff
@@ -7,6 +7,5 @@ option go_package = "github.com/palomachain/paloma/x/evm/types";
 
 // Params defines the parameters for the module.
 message Params {
-  option (gogoproto.goproto_stringer) = false;
```
Comment (byte-bandit) [inline proto/palomachain/paloma/evm/params.proto:10]: Why remove this?

File: x/evm/types/msgs.go [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+func (m *MsgUpdateParams) ValidateBasic() error {
+	if _, err := sdk.AccAddressFromBech32(m.Authority); err != nil {
+		return sdkerrors.Wrap(err, "invalid authority address")
+	}
+
+	if err := m.Params.Validate(); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (byte-bandit) [inline x/evm/types/msgs.go:31]: Take a look at the `ValidateBasic` and `GetSigners` implementation for other message types. Please make sure you follow the implementation of those instead of this.

Also make sure any new messages added embed the necessary metadata fields.

File: x/evm/types/params.go [hunk 18]
```diff
@@ -17,10 +17,10 @@ func NewParams() Params {
...
 }
 
-// DefaultParams returns a default set of parameters
-func DefaultParams() Params {
-	return NewParams()
-}
+// // DefaultParams returns a default set of parameters
+// func DefaultParams() Params {
+// 	return NewParams()
+// }
```
Comment (byte-bandit) [inline x/evm/types/params.go:23]: No commented out code.

File: x/evm/types/params_legacy.go [hunk 19]
```diff
@@ -0,0 +1 @@
+package types
```
Comment (byte-bandit) [inline x/evm/types/params_legacy.go:1]: Is this file needed?

**Review by byte-bandit — CHANGES_REQUESTED** [review]

File: proto/palomachain/paloma/evm/params.proto (hunk 16 above)
Comment (vishal-kanna) [reply proto/palomachain/paloma/evm/params.proto:10]: @byte-bandit  i think in v0.50 they have removed it .

Comment (vishal-kanna) [reply proto/palomachain/paloma/evm/params.proto:10]: ok

File: util/liblog/liblog.go (hunk 15 above)
Comment (vishal-kanna) [reply util/liblog/liblog.go:25]: but if we don't implement it we are getting an error that Impl() function is not implemented.

File: proto/palomachain/paloma/evm/params.proto (hunk 16 above)
Comment (deepan95dev) [reply proto/palomachain/paloma/evm/params.proto:10]: > Looks like it's still used in the base repo https://github.com/cosmos/gogoproto
 
in upgradation docs [v0.50.1 upgrade](https://github.com/cosmos/cosmos-sdk/blob/v0.50.1/UPGRADING.md#stringer), its mentioned so.  Should we retain it or keep it deleted? @byte-bandit

Comment (byte-bandit) [reply proto/palomachain/paloma/evm/params.proto:10]: Safe to remove then I'd say.

File: x/evm/keeper/setup_test.go [hunk 20]
```diff
@@ -1,15 +1,17 @@
 package keeper
 
 import (
-	tmdb "github.com/cometbft/cometbft-db"
-	"github.com/cometbft/cometbft/libs/log"
+	tmdb "github.com/cosmos/cosmos-db"
+	// "github.com/cometbft/cometbft/libs/log"
```
Comment (aleem1314) [inline x/evm/keeper/setup_test.go:5]:
```suggestion
```

File: x/evm/types/params_legacy.go (hunk 19 above)
Comment (vishal-kanna) [reply x/evm/types/params_legacy.go:1]: Yes it is needed for migration

Comment (byte-bandit) [reply x/evm/types/params_legacy.go:1]: Can you add a comment to the file with an explanation for this? Thanks.

File: go.mod [hunk 21]
```diff
@@ -3,33 +3,34 @@ module github.com/palomachain/paloma
 go 1.21
 
 require (
-	cosmossdk.io/errors v1.0.0 // indirect
+	cosmossdk.io/errors v1.0.0
 	cosmossdk.io/math v1.2.0
-	github.com/CosmWasm/wasmd v0.50.0-rc.0
-	github.com/CosmWasm/wasmvm v1.5.0 // indirect
+	// github.com/CosmWasm/wasmd v0.45.0
```
Comment (aleem1314) [inline go.mod:8]: Why wasm dependencies commented?

File: x/evm/keeper/keeper.go [hunk 22]
```diff
@@ -8,13 +8,15 @@ import (
 	"strings"
 	"time"
 
+	sdkmath "cosmossdk.io/math"
+	"cosmossdk.io/store/prefix"
+	storetypes "cosmossdk.io/store/types"
 	"github.com/VolumeFi/whoops"
-	"github.com/cometbft/cometbft/libs/log"
+
+	// "github.com/cometbft/cometbft/libs/log"
```
Comment (aleem1314) [inline x/evm/keeper/keeper.go:16]:
```suggestion
```

File: x/evm/module.go (hunk 9 above)
Comment (aleem1314) [inline x/evm/module.go:24]: ditto here

File: x/evm/types/keys.go [hunk 23]
```diff
@@ -1,5 +1,9 @@
 package types
 
+var (
+	ParamsKey = []byte{0x00}
```
Comment (aleem1314) [inline x/evm/types/keys.go:4]: Where are we using this?

File: x/evm/keeper/attest.go [hunk 24]
```diff
@@ -203,37 +204,37 @@ func (k Keeper) attestRouter(ctx sdk.Context, q consensus.Queuer, msg consensust
...
 			if err != nil {
-				logger.WithError(err).Error("Failed to get TX")
+				logger.With(err).Error("Failed to get TX")
 				return err
 			}
 			if k.isTxProcessed(ctx, tx) {
 				// somebody submitted the old transaction that was already processed?
 				// punish those validators!!
-				logger.WithError(err).Error("TX already processed")
+				logger.With(err).Error("TX already processed")
```
Comment (aleem1314) [inline x/evm/keeper/attest.go:213]: This is not resolved yet. Please revert these changes

File: x/evm/keeper/msg_assigner_test.go [hunk 25]
```diff
@@ -6,7 +6,9 @@ import (
 	"math"
 	"testing"
 
-	"github.com/cometbft/cometbft/libs/log"
+	// "github.com/cometbft/cometbft/libs/log"
```
[the remark this answers is not in the document]
Comment (vishal-kanna) [reply x/evm/keeper/msg_assigner_test.go:9]: ok

File: x/evm/types/keys.go (hunk 23 above)
Comment (vishal-kanna) [reply x/evm/types/keys.go:4]: Previously i thought that evm module had params.So that's why i haved added it .I'll remove it

File: x/evm/keeper/attest.go (hunk 24 above)
Comment (vishal-kanna) [reply x/evm/keeper/attest.go:213]: sure

**Comment by aleem1314** [comment]
Update this branch with `v0.50.0-upgrade` and revert all logger changes

File: x/evm/types/params.go [hunk 26]
```diff
@@ -1,39 +1,24 @@
...
-// DefaultParams returns a default set of parameters
-func DefaultParams() Params {
-	return NewParams()
-}
-
 // ParamSetPairs get the params.ParamSet
-func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
-	return paramtypes.ParamSetPairs{}
+func (p *Params) ParamSetPairs() ParamSetPairs {
+	return ParamSetPairs{}
```
Comment (aleem1314) [inline x/evm/types/params.go:15]: revert this change

File: go.mod [hunk 27]
```diff
@@ -3,33 +3,34 @@ module github.com/palomachain/paloma
 go 1.21
 
 require (
-	cosmossdk.io/errors v1.0.0 // indirect
+	cosmossdk.io/errors v1.0.0
 	cosmossdk.io/math v1.2.0
-	github.com/CosmWasm/wasmd v0.50.0-rc.0
-	github.com/CosmWasm/wasmvm v1.5.0 // indirect
+	github.com/CosmWasm/wasmd v0.45.0
```
Comment (aleem1314) [inline go.mod:8]: why wasm version changed in this PR?

File: app/app.go [hunk 28]
```diff
@@ -587,14 +583,14 @@ func New(
...
-	// 	appCodec,
-	// 	keys[evmmoduletypes.StoreKey],
-	// 	memKeys[evmmoduletypes.MemStoreKey],
-	// 	app.GetSubspace(evmmoduletypes.ModuleName),
-	// 	app.ConsensusKeeper,
-	// 	app.ValsetKeeper,
-	// )
+	app.EvmKeeper = *evmmodulekeeper.NewKeeper(
+		appCodec,
+		keys[evmmoduletypes.StoreKey],
```
Comment (akhilkumarpilli) [inline app/app.go:588]: Use `runtime.NewKVStoreService`  instead of StoreKey and update keeper to use StoreService
ref: https://github.com/cosmos/cosmos-sdk/blob/v0.50.1/UPGRADING.md#module-wiring

File: x/evm/keeper/keeper.go [hunk 29]
```diff
@@ -91,37 +92,30 @@ func init() {
...
 
 type Keeper struct {
-	cdc        codec.BinaryCodec
-	storeKey   storetypes.StoreKey
-	memKey     storetypes.StoreKey
-	paramstore paramtypes.Subspace
-
+	cdc             codec.BinaryCodec
+	storeKey        storetypes.StoreKey
+	memKey          storetypes.StoreKey
```
Comment (akhilkumarpilli) [inline x/evm/keeper/keeper.go:97]: ditto, update to use `KVStoreService`

File: x/evm/types/params.go (hunk 26 above)
Comment (deepan95dev) [reply x/evm/types/params.go:15]: previously it was using `paramtypes` from cosmos-sdk `x/params` and this folder is not being used anywhere in 0.50.1 so the required types are moved to local param types and `x/params` paramtypes is removed

File: x/evm/keeper/keeper.go (hunk 29 above)
Comment (vishal-kanna) [reply x/evm/keeper/keeper.go:98]: Done

File: app/app.go [hunk 30]
```diff
@@ -587,14 +583,14 @@ func New(
...
-	// 	app.GetSubspace(evmmoduletypes.ModuleName),
-	// 	app.ConsensusKeeper,
-	// 	app.ValsetKeeper,
-	// )
+	app.EvmKeeper = *evmmodulekeeper.NewKeeper(
+		appCodec,
+		runtime.NewKVStoreService(keys[evmmoduletypes.StoreKey]),
+		memKeys[evmmoduletypes.MemStoreKey],
+		nil,
+		nil,
```
Comment (akhilkumarpilli) [inline app/app.go:590]: Why these values are nil? Remove these fields if not needed or add TODO comment if they need to be updated later.

**Review by akhilkumarpilli — COMMENTED** [review]
Still there are few things need to be updated:
* Add signer field proto annotation, remove `GetSigners` and `GetSignBytes` implementations in every sdk.Msg defined in this module.
* Add EndBlocker, BeginBlocker and other compiler assertions (if required any) in module.go.
* Update Endblocker method structure to new SDK defined type if there are not validator updates involved in it.
* Update BeginBlocker method structure.

Please go through [migration docs](https://github.com/cosmos/cosmos-sdk/blob/v0.50.1/UPGRADING.md#v050x) much deeper and update any if still not updated.

**Comment by byte-bandit** [comment]
> Still there are few things need to be updated:
> 
> * Add signer field proto annotation, remove `GetSigners` and `GetSignBytes` implementations in every sdk.Msg defined in this module.

> [!WARNING] 
> Paloma is using a custom signer function using an embedded `Metadata` field on every message. There's a custom Ante handler responsible for parsing and verifying signatures on messages. This should stay as is and not be touched. `GetSigners` and `GetSignBytes` should be handled by `libmeta` on every message, but they need to stay implemented for this to work.

File: proto/palomachain/paloma/evm/tx.proto [hunk 31]
```diff
@@ -2,8 +2,10 @@ syntax = "proto3";
 package palomachain.paloma.evm;
 
 import "gogoproto/gogo.proto";
+import "cosmos_proto/cosmos.proto";
+import "cosmos/msg/v1/msg.proto";
 import "palomachain/paloma/valset/common.proto";
-
+import "palomachain/paloma/evm/params.proto";
```
Comment (aleem1314) [inline proto/palomachain/paloma/evm/tx.proto:8]:
```suggestion
```

File: x/evm/keeper/attest.go [hunk 32]
```diff
@@ -58,13 +61,14 @@ func (c *consensusPower) consensus() bool {
...
+	return c.runningSum.Mul(sdkmath.NewInt(3)).GTE(
+		c.totalPower.Mul(sdkmath.NewInt(2)),
 	)
 }
 
-func (k Keeper) attestRouter(ctx sdk.Context, q consensus.Queuer, msg consensustypes.QueuedSignedMessageI) (err error) {
-	logger := k.Logger(ctx).WithFields(
+func (k Keeper) attestRouter(ctx context.Context, q consensus.Queuer, msg consensustypes.QueuedSignedMessageI) (err error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	logger := k.Logger(sdkCtx).WithFields(
```
Comment (byte-bandit) [inline x/evm/keeper/attest.go:71]: @aleem1314 @vishal-kanna 

Please start using `liblog.FromSDKLogger(k.Logger(SdkCtx)` in the future. There are still ocurrances here without.
